### PR TITLE
blocks-components-tests

### PR DESCRIPTION
**Pull request by FarukOzderim**
#894 
- move input and output tests under test_components

Did not delve into individual tests that much, WDYT about these tests quality @abidlabs? Would we want increase quality in these tests?

### CODE REVIEW

**Comment by abidlabs** [comment]
Yes, I think we should increase the "quality" of the tests. Based on our conversation, specifically:
* Making the tests more readable by organizing them by component methods. So for example, each `TestComponentX` class would have a `test_processing()` method, a `test_save_flagged()` method, a `test_tokenize()` method, etc.
* In addition to these methods, having a clearly separate "component end-to-end test", which tests the complete flow of preprocessing -> function -> postprocessing for each component. These tests should be written in a way that they illustrate the various ways that a component can be used. For example, for `Image`, we should have tests that show how images can be accepted as PIL objects, numpy arrays, or filepaths. 
* Avoiding the use of `Interface.launch()` whenever possible (I think we do a pretty good job of this)

In general, my biggest concern is the lack of readability of the tests. For example, I see a test like this and it is far from apparent what this is testing:

```python
    def test_in_interface(self):
        iface = gr.Interface(lambda x: x**2, "number", "textbox")
        self.assertEqual(iface.process([2])[0], ["4.0"])
        iface = gr.Interface(
            lambda x: x**2, "number", "number", interpretation="default"
        )
        scores, alternative_outputs = iface.interpret([2])
        self.assertEqual(
            scores,
            [
                [
                    (1.94, -0.23640000000000017),
                    (1.96, -0.15840000000000032),
                    (1.98, -0.07960000000000012),
                    [2, None],
                    (2.02, 0.08040000000000003),
                    (2.04, 0.16159999999999997),
                    (2.06, 0.24359999999999982),
                ]
            ],
        )
        self.assertEqual(
            alternative_outputs,
            [
                [
                    [3.7636],
                    [3.8415999999999997],
                    [3.9204],
                    [4.0804],
                    [4.1616],
                    [4.2436],
                ]
            ],
        )
```

I think we should make these changes in this PR, but let me know if you wanna do it separately.

**Comment by FarukOzderim** [comment]
One of our problem in tests is not having idempotency, for example after I added the file `test_components` the test below started to fail, and it just checks the preprocessed constant's shape, I don't understand why it fails. Will fix it by having another copy of the constant.

I don't get the problem with not having idempotency situations, we will need to investigate further to understand the problem.
```

TestAudio.test_as_component __________________________

self = <test.test_inputs.TestAudio testMethod=test_as_component>

    def test_as_component(self):
        x_wav = gr.test_data.BASE64_AUDIO
        audio_input = gr.inputs.Audio()
        output = audio_input.preprocess(x_wav)
        self.assertEqual(output[0], 8000)
>       self.assertEqual(output[1].shape, (8046,))
E       AssertionError: Tuples differ: (241,) != (8046,)
E       
E       First differing element 0:
E       241
E       8046
E       
E       - (241,)
E       + (8046,)

test/test_inputs.py:514: AssertionError

```

File: test/test_processing_utils.py [hunk 1]
```diff
@@ -13,41 +14,41 @@
...
         )
-        self.assertEqual(output_base64, gr.test_data.BASE64_IMAGE)
+        self.assertEqual(output_base64, media_data.BASE64_IMAGE)
 
     # Commented out because this is throwing errors on Windows. Possibly due to different matplotlib behavior on Windows?
     # def test_encode_plot_to_base64(self):
     #     plt.plot([1, 2, 3, 4])
     #     output_base64 = gr.processing_utils.encode_plot_to_base64(plt)
-    #     self.assertEqual(output_base64, gr.test_data.BASE64_PLT_IMG)
+    #     self.assertEqual(output_base64, media_data.BASE64_PLT_IMG)
```
Comment (FarukOzderim) [inline test/test_processing_utils.py:44]: @abidlabs I am commenting out this test while fixing the expected output value, it does not match.
```

E   AssertionError: 'data[107 chars]24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK[23085 chars]QmCC' != 'data[107 chars]24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkT[23085 chars]QmCC'
E   Diff is 46821 characters long. Set self.maxDiff to None to see it.
```

**Comment by FarukOzderim** [comment]
@abidlabs I added restore_flagging to Carousel component, but it gives an error, don't we support deserialize in Carousel component?

```
        if isinstance(s, str):
            if s.startswith('\ufeff'):
                raise JSONDecodeError("Unexpected UTF-8 BOM (decode using utf-8-sig)",
                                      s, 0)
        else:
            if not isinstance(s, (bytes, bytearray)):
>               raise TypeError(f'the JSON object must be str, bytes or bytearray, '
                                f'not {s.__class__.__name__}')
E               TypeError: the JSON object must be str, bytes or bytearray, not list

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2800.0_x64__qbz5n2kfra8p0\lib\json\__init__.py:339: TypeError

```

**Comment by FarukOzderim** [comment]
Another strange error that occurs in circleci that I have not met on my machine is below. It seems there is an issue with ffmpy installation.

https://app.circleci.com/pipelines/github/gradio-app/gradio/2231/workflows/b9386f6a-85f7-47f3-a104-f872560a359f/jobs/2258?invite=true#step-114-457

```
ffmpy.FFExecutableNotFoundError: Executable 'ffmpeg' not found

venv/lib/python3.9/site-packages/ffmpy.py:98: FFExecutableNotFoundError

```

**Comment by FarukOzderim** [comment]
It is strange because ffmpy's last release was on Jan 13 2021
https://pypi.org/search/?q=ffmpy

**Comment by FarukOzderim** [comment]
Now unit tests are covering essential functions within components. There are a few functional tests that are missing(labeled with TODO), either you or I(with your help) can work on them @abidlabs. Also if we see a need for more complex or in depth functional test flows we can add them as well.

File: test/test_blocks.py [hunk 2]
```diff
@@ -1,10 +1,6 @@
 import random
 import unittest
-
-try:
-    from .test_data.blocks_configs import XRAY_CONFIG  # for pytest
-except ImportError:
-    from test_data.blocks_configs import XRAY_CONFIG  # for regular python
+from test.test_data.blocks_configs import XRAY_CONFIG
```
Comment (abidlabs) [inline test/test_blocks.py:3]: Can we revert this change? Without this, I get import errors:

```
ModuleNotFoundError: No module named 'test.test_data'
```

File: test/test_components.py [hunk 3]
```diff
@@ -0,0 +1,1496 @@
...
+import pandas as pd
+import PIL
+
+import gradio as gr
+
+os.environ["GRADIO_ANALYTICS_ENABLED"] = "False"
+
+"""
+Tests are divided into two
+1. test_component_functionalities are unit tests that check essential functions of a component, the functions that are checked are documented in the docstring.
```
[the remark this answers is not in the document]
Comment (abidlabs) [reply test/test_components.py:20]: I think its might. But functionality is a weird word, I might suggest using: `test_component_functions`

File: test/test_blocks.py (hunk 2 above)
Comment (FarukOzderim) [reply test/test_blocks.py:3]: Can you try `python -m pytest test/test_blocks.py`? Or run the whole tests with `bash scripts/run_all_tests.sh`.
If possible I would like to stay away from this syntax.

**Comment by abidlabs** [comment]
This is great @FarukOzderim!

I made some tweaks to get the tests in line with @aliabid94's changes to the interpretation and fixed the carousel tests.

Not sure about the `ffmpeg` error, that's weird because I thought we tested video before as well. Anyways perhaps we need to install ffmpeg on the circleci machines. Can you take a look at this and see if it resolves the issue? https://linuxize.com/post/how-to-install-ffmpeg-on-ubuntu-18-04/

File: test/test_components.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (FarukOzderim) [reply test/test_components.py]: Done

**Comment by FarukOzderim** [comment]
> This is great @FarukOzderim!
> 
> I made some tweaks to get the tests in line with @aliabid94's changes to the interpretation and fixed the carousel tests.
> 
> Not sure about the `ffmpeg` error, that's weird because I thought we tested video before as well. Anyways perhaps we need to install ffmpeg on the circleci machines. Can you take a look at this and see if it resolves the issue? https://linuxize.com/post/how-to-install-ffmpeg-on-ubuntu-18-04/

Well, the reason I think it is not about ffmpeg being installed independently is that there was no problem on other PR's tests.
ie: https://app.circleci.com/pipelines/github/gradio-app/gradio/2232/workflows/17d1f661-21c7-4902-a760-23594b663c0d/jobs/2259

**Comment by abidlabs** [comment]
> Let me try nonetheless. Though if we need to install it independently that means user will need to as well right?

I'm pretty sure most systems have ffmpeg already installed. Maybe circle CI machines don't. Tbh I'm not sure

**Comment by abidlabs** [comment]
Looks like it worked! LGTM
